Fixes field validation issues #2179 and #1963

blitux opened the pull request:
Hi there. This PR fixes both #2179 and #1963. 

The file app.js needed some small refactoring to handle some special conditions (like field names with brackets). The publishable app.js was compiled with production mode. 

Comments are welcome. Thanks!

samtheson commented:
Some notes from a fix I had prepared (preferring yours so that if the refactor is accepted we get all the good stuff in one):

- You left the original PR's changes. We need to clean them up https://github.com/samtheson/voyager/commit/c10e3546ff1de4fac0a0361f0c6dd528e5b01d7b
- You removed the original PR's added functionality, which is cool and needed. You can go with the same approach to that in your `getFieldsWithValidationRules()`, my implementation of the same functionality in the non-refactored way is at https://github.com/samtheson/voyager/commit/4674dea93ae00345d0b08b2995cb2093071cdc2e

blitux commented:
Excellent! I can make another PR for the feature If you want. I prefer to have features and bugfixes separated to make them as atomic as possible. I helps tracking the changes and can be reverted easily if something goes wrong.

samtheson commented:
Agreed.

Another PR will do, I think. Because this PR here solves the bug but removes a feature from Voyager.

blitux commented:
Sorry, didn't see that. The feature of using custom attributes to improve validation messages? Was that a feature that existed before or is a feature you were adding on your PR? If is the latter, I can add it on another different PR because it's a new feature. But if not, I'l be glad to add it on the current PR. 

Also, let's try to make it to the release ASAP as this is a critical issue and many people may be blocked by this. @fletch3555 who is in charge of merging PRs?

samtheson commented:
No worries :+1:, still to come to full grip with the codebase myself. Just make sure to also include a way to have the better validation messages. The PR so far has the functionality removed, so a release will only end up in a new kind of issue.

fletch3555 commented on resources/assets/js/app.js:
```diff
@@ -134,30 +133,34 @@ $(document).ready(function(){
-        $("div").removeClass("has-error");
-        $(".help-block").remove();
+      beforeSend: function () {
+        $('body').css('cursor', 'progress');
```
Just out of curiosity, any reason for the double -> single quote change?  Not sure if there was some intent other than "appease StyleCI"

blitux replied:
No, It was just a personal preference. I use double quotes for strings where I do interpolation and single quotes for plain strings. But that's on PHP's side. But on javascript they behave the same AFAIK.

fletch3555 commented on resources/assets/js/app.js:
```diff
@@ -134,30 +133,34 @@ $(document).ready(function(){
+        $('body').css('cursor', 'auto');
+        $.each(d.errors, function (inputName, errorMessage) {
+          // Using ^ 'starts with' operator to also look for input arrays
+          var $inputElement = $("[name^='" + inputName + "']"),
```
My only hesitation with the 'starts with' operator is the case of `address` vs `address2`.

blitux replied:
Yes, In this case it will return a collection of objects, and then get the first. But just realized that the first element of the collection may not be the same object I'm looking for... Will try to come up with some better solution for this.

fletch3555 commented:
So are we saying this PR is removing existing functionality, or just not including functionality that PR #2190 did?  If the former, it should really be added back in (though I don't see what it would be on the changes tab).  If the latter, I agree that can be a separate "new feature" PR

samtheson commented:
@fletch3555 This PR solves the issue by _removing_ functionality. We lose the functionality introduced in PR #2077. I had favoured this one over mine because Pablo had some ideas I thought I should give preference.

But since this PR is more of a refactor and needs some work, I submitted a PR which reverses PR #2077 (solves bug), keeps our old working code as is (no refactor) and implements the functionality in a better way (maintains functionality) :point_right:  https://github.com/the-control-group/voyager/pull/2198

So that's where Pablo & I were having a bit of miscommunicating earlier on Slack, he hadn't yet (before I properly explained) seen the functionality the problematic PR had introduced and so it looked like my suggestions were _adding_ new functionality to Voyager, and his solution thus ended up fixing the issue by removing functionality.

@blitux Very much looking forward to the final polish off of this PR though. New ideas excite me :)

blitux commented:
@samtheson I think your aproach of custom attributes is ok. Now I understand what it was for. Will add it on the same PR because that was the intended functionality from the beginning, but was not executed properly.

blitux commented:
@samtheson @fletch3555 hey guys, do you mind to do a quick review? I think this is the last revision of the PR. I did some quick testing and the validation errors are being shown properly, including the field display name for the fields that have it. And also is handling correctly the case of brackets in the name of the field. I tested it with the multiple image[] field.

samtheson commented:
Will do. Just going to go through all [QA Needed](https://github.com/the-control-group/voyager/labels/QA%20Needed) this morning.

blitux commented:
Hi guys, any news on this?

blitux commented:
Maintainers: what is needed for this to be merged? It's a critical issue (form validation is broken and it's not possible to save records from BREAD forms). I've done a QA Test starting from a new Voyager installation and it's working. I would need a confirmation (as I've sent the PR and did the test by myself). This has been waiting for review since nov 8.

judirickonto commented:
thnk @blitux sir, it helps alot.

blitux commented:
Hello guys, 

@fletch3555 I've rebased the PR with 1.0 (which is now our 'master' branch) and fixed the conflicts.

Can someone review this pleaaaase?

blitux commented:
Updated PR to the last revision.

blitux commented:
I'm closing this in favour of a new PR I'll send.
